cmd/singularity/instance_test.go

```diff
@@ -6,67 +6,420 @@
 package main
 
 import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
 	"os"
 	"os/exec"
+	"path/filepath"
+	"strconv"
 	"testing"
 
 	"github.com/sylabs/singularity/internal/pkg/test"
 )
 
-func startInstance(image string, instance string) ([]byte, error) {
-	args := []string{"instance", "start", image, instance}
-	cmd := exec.Command(cmdPath, args...)
+const instanceStartPort = 11372
+const instanceDefinition = "../../examples/instances/Singularity"
+const instanceImagePath = "./instance_tests.sif"
+
+type startOpts struct {
+	addCaps       string
+	allowSetuid   bool
+	applyCgroups  string
+	bind          string
+	boot          bool
+	cleanenv      bool
+	contain       bool
+	containall    bool
+	dns           string
+	dropCaps      string
+	home          string
+	hostname      string
+	keepPrivs     bool
+	net           bool
+	network       string
+	networkArgs   string
+	noHome        bool
+	noPrivs       bool
+	nv            bool
+	overlay       string
+	scratch       string
+	security      string
+	userns        bool
+	uts           bool
+	workdir       string
+	writable      bool
+	writableTmpfs bool
+}
+
+type listOpts struct {
+	json      bool
+	user      string
+	container string
+}
+
+type stopOpts struct {
+	all      bool
+	force    bool
+	signal   string
+	timeout  string
+	user     string
+	instance string
+}
 
+type instance struct {
+	Instance string `json:"instance"`
+	Pid      int    `json:"pid"`
+	Image    string `json:"img"`
+}
+
+type instanceList struct {
+	Instances []instance `json:"instances"`
+}
+
+func startInstance(image string, instance string, opts startOpts) ([]byte, error) {
+	args := []string{"instance", "start"}
+	if opts.addCaps != "" {
+		args = append(args, "--add-caps", opts.addCaps)
+	}
+	if opts.allowSetuid {
+		args = append(args, "--allow-setuid")
+	}
+	if opts.applyCgroups != "" {
+		args = append(args, "--apply-cgroups", opts.applyCgroups)
+	}
+	if opts.bind != "" {
+		args = append(args, "--bind", opts.bind)
```

Comment:
Check https://github.com/sylabs/singularity/blob/master/cmd/singularity/actions_test.go#L44 ,so we can specify multiple binds on a test (not know but in a near future)

Reply:
So I actually did look into that, but I'm a bit confused by something. The docs say that one bind accepts a single, comma-delimited string, but there's no indication that multiple `--bind`s can be supplied.
```
  -B, --bind strings           a user-bind path specification.  spec has
                               the format src[:dest[:opts]], where src and
                               dest are outside and inside paths.  If dest
                               is not given, it is set equal to src.
                               Mount options ('opts') may be specified as
                               'ro' (read-only) or 'rw' (read/write, which
                               is the default). Multiple bind paths can be
                               given by a comma separated list.
```

Reply:
Look that the `for `loop appends a `--bind` along with the bind string
```go
argv = append(argv, "--bind", bind)
```

Reply:
Fixed; thanks!